### PR TITLE
NAS-123910 / 24.04 / Roles

### PR DESCRIPTION
**Pull request by themylogin**
We would like to have 3 different roles in API to begin with:

- Full Admin: just like root/admin
- Read-only: Access to all readonly endpoints (e.g. query, choices, get, etc). No access to any data changing methods (create, update, delete)
- Shares Manager: Be able to create/edit datasets and shares. No be able to do any other system admin tasks nor delete/destroy anything.

More information is in the ticket.

### CODE REVIEW

**Comment by themylogin** [comment]
time 4:00

**Comment by themylogin** [comment]
@bmeagherix @anodos325 ping

**Comment by anodos325** [comment]
Does this mean that a readonly user will have read access to the user DB table?

**Review by anodos325 — CHANGES_REQUESTED** [review]
We should have some handling for private fields in READONLY role otherwise there is potential for privilege escalation / unintended disclosure of information.

**Review by bmeagherix — CHANGES_REQUESTED** [review]
What is there looks very good, but I agree with Andrew's comment:
> We should have some handling for private fields in READONLY role otherwise there is potential for privilege escalation / unintended disclosure of information.

(For example, once we add a READ role covering it, I could envisage wanting to redact CHAP credentials from `iscsi.auth.query`.)

Have we worked out yet how the webui will behave for a user with READONLY role?

Wrt tests, IMO it would be good to add success/failure method calls in nfs, smb & iscsi for SHARING_READ role.  Also success calls (using the same methods) for SHARING_WRITE role.

**Comment by yocalebo** [comment]
After some discussion, hiding `private=True` fields from `query` endpoints is a huge non-trivial change that needs to happen in another PR. This PR just adds the exact items that were requested from management so we're going to hold off on hiding sensitive fields from query endpoints. I will open a separate ticket and link it to this one to implement said feature. For now, let's forgo that thought process and get another round of reviews please @anodos325 @bmeagherix

**Review by bmeagherix — CHANGES_REQUESTED** [review]
Suggest adding tests for success/failure method calls in nfs, smb & iscsi for SHARING_READ role.

Also success calls (using the same methods) for SHARING_WRITE role.

**Comment by anodos325** [comment]
I think we should add privilege table entries for each of the roles that don't already exist (we already have an admin one with full control).

**Comment by anodos325** [comment]
I think we should also expand the dump of logged-in user `auth.me` to include information about privileges assigned to user (based on group membership).

**Comment by themylogin** [comment]
> I think we should add privilege table entries for each of the roles that don't already exist (we already have an admin one with full control).

@yocalebo what do you think of this? Looks questionable, we'll have a ton of useless entries by default.

**Comment by anodos325** [comment]
> > I think we should add privilege table entries for each of the roles that don't already exist (we already have an admin one with full control).
> 
> @yocalebo what do you think of this? Looks questionable, we'll have a ton of useless entries by default.

Well, maybe not all of them, but one or two that cover some basic common use-cases (like the READONLY access). In that way it can kind of act as template for users to think "oh this is how it's done". Just a thought.

**Comment by themylogin** [comment]
I agree on adding "readonly user" built-in privilege. @yocalebo ?

**Comment by yocalebo** [comment]
@themylogin yeah adding read only is a good first step for now

File: src/middlewared/middlewared/alembic/versions/24.04/2023-10-30_21-02_readonly_privilege.py
Comment (anodos325) [inline src/middlewared/middlewared/alembic/versions/24.04/2023-10-30_21-02_readonly_privilege.py:23]: I believe per our previous conversations about sensitive info being exposed to READONLY users that we may want to explicitly name something like "READONLY ADMINISTRATORS", correct?

**Review by yocalebo — APPROVED** [review]
Great job but please wait for @anodos325 and @bmeagherix approval before merge.

**Review by anodos325 — APPROVED** [review]

**Comment by themylogin** [comment]
time 8:00

**Pull request merged**
